client/taskenv/services.go

```diff
@@ -31,42 +31,174 @@ func InterpolateServices(taskEnv *TaskEnv, services []*structs.Service) []*struc
 			check.InitialStatus = taskEnv.ReplaceEnv(check.InitialStatus)
 			check.Method = taskEnv.ReplaceEnv(check.Method)
 			check.GRPCService = taskEnv.ReplaceEnv(check.GRPCService)
-			if len(check.Header) > 0 {
-				header := make(map[string][]string, len(check.Header))
-				for k, vs := range check.Header {
-					newVals := make([]string, len(vs))
-					for i, v := range vs {
-						newVals[i] = taskEnv.ReplaceEnv(v)
-					}
-					header[taskEnv.ReplaceEnv(k)] = newVals
-				}
-				check.Header = header
-			}
+			check.Header = interpolateMapStringSliceString(taskEnv, check.Header)
 		}
 
 		service.Name = taskEnv.ReplaceEnv(service.Name)
 		service.PortLabel = taskEnv.ReplaceEnv(service.PortLabel)
 		service.Tags = taskEnv.ParseAndReplace(service.Tags)
 		service.CanaryTags = taskEnv.ParseAndReplace(service.CanaryTags)
+		service.Meta = interpolateMapStringString(taskEnv, service.Meta)
+		service.CanaryMeta = interpolateMapStringString(taskEnv, service.CanaryMeta)
+		service.Connect = interpolateConnect(taskEnv, service.Connect)
```

Comment:
Somewhat related because we're potentially setting these to `nil` and not empty maps, none of these three fields is [`Canonicalized`](https://github.com/hashicorp/nomad/blob/v1.0.0/nomad/structs/services.go#L448-L473). Should we be doing that and/or should we be assigning an empty map here for safety?

Reply:
The canonicalized form of empty maps is `nil` (at least in services + consul)
https://github.com/hashicorp/nomad/blob/a480eed0815c54612856d9115a34bb1d1a773e8c/nomad/structs/services.go#L174

`Copy` also make use of the helper.Copy* functions which also reduce to `nil` if there are no elements 
https://github.com/hashicorp/nomad/blob/master/helper/funcs.go#L282